### PR TITLE
DI-191 Configure API Gateway endpoint with route53 record

### PR DESCRIPTION
**Pull request by rohith1122**
## Link to JIRA Ticket

- https://nhsd-jira.digital.nhs.uk/browse/DI-191

## Description
This ticket  adds the terraform script to allocate  DNS name to the existing APIGateway endpoint.

### Noteworthy Changes
Creates 3 resources and uses 3 existing data sources  in AWS from this script 
-   Route53 Zone (data source)
-   ACM Certificate (data source)
-   Rest API Name (data source)
-   Configure subdomain
-   API Gateway path mapping 
-   Route53 A record for API Gateway 

## Testing

Please tick the testing that has been completed

- [ ] Unit tests
- [ ] Component tests

## Developer Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have run the [code formatting checks](../README.md#code-quality)
- [ ] I have run the [code quality checks](../README.md#code-quality)
- [ ] New code meets [standards](https://nhsd-confluence.digital.nhs.uk/display/DI/DI+Ways+of+Working) agreed by the team
- [ ] Unit test code coverage is at or above 80%
- [ ] New and existing unit tests pass locally with my changes
- [ ] Tests have added that prove my fix is effective or that my feature works (Component tests)
- [ ] I have made corresponding changes to the documentation
- [ ] I have cleaned down my environment (if created)

## Code Reviewer Checklist

- [ ] I have run the unit tests and they run correctly
- [ ] I can confirm the changes have been tested or approved by a tester
- [ ] I can confirm no remaining infrastructure is left over from this branch

### CODE REVIEW

File: infrastructure/stacks/api-gateway-route53/main.tf [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+resource "aws_api_gateway_base_path_mapping" "uec-dos-integration_api_mapping" {
+  api_id      = data.aws_api_gateway_rest_api.restapi.id
+  stage_name  = var.profile
+  domain_name = "${var.dos_integration_sub_domain_name}.${var.terraform_networking_route53_zone_name}"
+}
+
+resource "aws_route53_record" "uec-dos-integration_api_endpoint" {
+  name    = "${var.dos_integration_sub_domain_name}.${var.terraform_networking_route53_zone_name}"
+  type    = "A"
```
Comment (JackPlowman) [inline infrastructure/stacks/api-gateway-route53/main.tf:18]: Should this be a CNAME?

File: infrastructure/stacks/api-gateway-route53/main.tf [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  zone_id = data.aws_route53_zone.k8s_route53_zone.zone_id
```
Comment (JackPlowman) [inline infrastructure/stacks/api-gateway-route53/main.tf:19]: This needs to be `data.terraform_remote_state.route53.outputs.dns_zone_id`

**Review by JackPlowman — CHANGES_REQUESTED** [review]
This is the first step towards getting the terraform to work. There will be more changes at a later stage.

**Comment by JackPlowman** [comment]
It wouldn't let me put these in the review but can you add these variables in `/build/automation/var/project.mk`
```
TF_VAR_team_id = $(TEAM_ID)
TF_VAR_programme = $(PROGRAMME)
```

It wouldn't let me put these in the review but can you add these variables in `/build/automation/var/profile/task.mk`. Once these variables have been confirmed they will have to be copied to dev.mk, demo.mk and live.mk
```
# API Gateway Route53
TF_VAR_dos_integration_sub_domain_name := $(PROJECT_ID)-$(ENVIRONMENT)
```

File: infrastructure/stacks/api-gateway-route53/variables.tf [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+variable "terraform_platform_state_store" {
+  description = ""
+}
+
+variable "vpc_terraform_state_key" {
+  description = ""
+}
+
+
```
[the remark this answers is not in the document]
Comment (rohith1122) [reply infrastructure/stacks/api-gateway-route53/variables.tf:54]: Done

File: infrastructure/stacks/api-gateway-route53/main.tf (hunk 1 above)
Comment (rohith1122) [reply infrastructure/stacks/api-gateway-route53/main.tf]: Updated

File: infrastructure/stacks/api-gateway-route53/main.tf (hunk 2 above)
Comment (rohith1122) [reply infrastructure/stacks/api-gateway-route53/main.tf]: Updated

File: infrastructure/stacks/api-gateway-route53/main.tf [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# ############################
```
Comment (JackPlowman) [inline infrastructure/stacks/api-gateway-route53/main.tf:1]: Can this file be replaced to 
``` suggestion
# ############################
# # Route53 and APIGateway Domain
# ############################
resource "aws_api_gateway_domain_name" "api_gateway_domain_name" {
  regional_certificate_arn = data.aws_acm_certificate.issued.arn
  domain_name     = "${var.dos_integration_sub_domain_name}.${var.texas_hosted_zone}"
  security_policy = "TLS_1_2"
  endpoint_configuration {
    types = [ "REGIONAL" ]
  }
}

resource "aws_route53_record" "uec_dos_integration_api_endpoint" {
  name    = "${var.dos_integration_sub_domain_name}.${var.texas_hosted_zone}"
  type    = "A"
  zone_id = data.terraform_remote_state.route53.outputs.dns_zone_id
  alias {
    name                   = aws_api_gateway_domain_name.api_gateway_domain_name.regional_domain_name
    zone_id                = aws_api_gateway_domain_name.api_gateway_domain_name.regional_zone_id
    evaluate_target_health = true
  }
}

resource "aws_api_gateway_base_path_mapping" "uec-dos-integration_api_mapping" {
  api_id      = data.aws_api_gateway_rest_api.restapi.id
  stage_name  = var.environment
  domain_name = "${var.dos_integration_sub_domain_name}.${var.texas_hosted_zone}"
}

```

File: infrastructure/stacks/api-gateway-route53/variables.tf [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+# ############################
+# # ROUTE53
+# ############################
+
+
+variable "dos_integration_sub_domain_name" {
+  type        = string
+  description = "sub domain name"
+}
+
```
Comment (JackPlowman) [inline infrastructure/stacks/api-gateway-route53/variables.tf:64]: Please add this environment variable

``` suggestion 
variable "texas_hosted_zone" {
  description = ""
}
```

File: infrastructure/stacks/api-gateway-route53/variables.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (JackPlowman) [reply infrastructure/stacks/api-gateway-route53/variables.tf:54]: Please add this variable
``` suggestion
variable "environment" {
  description = ""
}
```

File: infrastructure/stacks/api-gateway-route53/data.tf [hunk 6]
```diff
@@ -0,0 +1,22 @@
+data "terraform_remote_state" "route53" {
```
Comment (JackPlowman) [inline infrastructure/stacks/api-gateway-route53/data.tf:1]: Can this file be modifed to
``` suggestion
data "terraform_remote_state" "route53" {
  backend = "s3"
  config = {
    bucket = var.terraform_platform_state_store
    key    = var.route53_terraform_state_key
    region = var.aws_region
  }
}

data "aws_acm_certificate" "issued" {
  domain   = "*.${var.texas_hosted_zone}"
  statuses = ["ISSUED"]
}

data "aws_api_gateway_rest_api" "restapi" {
  name = "${var.environment}-${var.programme}-${var.team_id}"
}
```

File: build/automation/var/profile/task.mk [hunk 7]
```diff
@@ -20,3 +20,6 @@ TF_VAR_nhs_uk_api_key_key := NHS_UK_API_KEY
 
 # Lambda Security Group
 TF_VAR_lambda_security_group_name := $(PROJECT_ID)-$(ENVIRONMENT)-lambda-sg
+
+# API Gateway Route53
```
Comment (JackPlowman) [inline build/automation/var/profile/task.mk:24]: Can this environment variable be modified to 
``` suggestion
# API Gateway Route53
TF_VAR_dos_integration_sub_domain_name := $(PROGRAMME)-$(TEAM_ID)-$(ENVIRONMENT)
DOS_INTEGRATION_URL := $(TF_VAR_dos_integration_sub_domain_name).$(TEXAS_HOSTED_ZONE)
```

File: build/automation/var/profile/live.mk [hunk 8]
```diff
@@ -1,3 +1,6 @@
 -include $(VAR_DIR)/platform-texas/v1/account-live-k8s-prod.mk
 
 LOG_LEVEL := INFO
+
+# API Gateway Route53
```
Comment (JackPlowman) [inline build/automation/var/profile/live.mk:5]: Can this environment variable be modified to 
``` suggestion
# API Gateway Route53
TF_VAR_dos_integration_sub_domain_name := $(PROGRAMME)-$(TEAM_ID)-$(ENVIRONMENT)
DOS_INTEGRATION_URL := $(TF_VAR_dos_integration_sub_domain_name).$(TEXAS_HOSTED_ZONE)
```

File: build/automation/var/profile/dev.mk [hunk 9]
```diff
@@ -21,3 +21,6 @@ TF_VAR_nhs_uk_api_key_key := NHS_UK_API_KEY
 
 # Lambda Security Group
 TF_VAR_lambda_security_group_name := $(PROJECT_ID)-$(ENVIRONMENT)-lambda-sg
+
+# API Gateway Route53
```
Comment (JackPlowman) [inline build/automation/var/profile/dev.mk:25]: Can this environment variable be modified to 
``` suggestion
# API Gateway Route53
TF_VAR_dos_integration_sub_domain_name := $(PROGRAMME)-$(TEAM_ID)-$(ENVIRONMENT)
DOS_INTEGRATION_URL := $(TF_VAR_dos_integration_sub_domain_name).$(TEXAS_HOSTED_ZONE)
```

File: build/automation/var/profile/demo.mk [hunk 10]
```diff
@@ -1,3 +1,6 @@
 -include $(VAR_DIR)/platform-texas/v1/account-live-k8s-prod.mk
 
 LOG_LEVEL := INFO
+
+# API Gateway Route53
```
Comment (JackPlowman) [inline build/automation/var/profile/demo.mk:5]: Can this environment variable be modified to 
``` suggestion
# API Gateway Route53
TF_VAR_dos_integration_sub_domain_name := $(PROGRAMME)-$(TEAM_ID)-$(ENVIRONMENT)
DOS_INTEGRATION_URL := $(TF_VAR_dos_integration_sub_domain_name).$(TEXAS_HOSTED_ZONE)
```

File: build/automation/var/project.mk [hunk 11]
```diff
@@ -24,3 +24,6 @@ DEPLOYMENT_SECRETS = $(PROJECT_ID)-$(PROFILE)/deployment
 AWS_SAME_ACCOUNT_DOCKER_REGISTRY = $(AWS_ACCOUNT_ID).dkr.ecr.$(AWS_DEFAULT_REGION).amazonaws.com/$(PROJECT_GROUP_SHORT)/$(PROJECT_NAME_SHORT)
 PYTHON_VERSION = $(PYTHON_VERSION_MAJOR).$(PYTHON_VERSION_MINOR)
 AWS_VPC_NAME = lk8s-$(AWS_ACCOUNT_NAME).texasplatform.uk
+
+TF_VAR_team_id = $(TEAM_ID)
+TF_VAR_programme = $(PROGRAMME)
```
Comment (JackPlowman) [inline build/automation/var/project.mk:29]: Can you add these two environment variables

``` suggestion
TF_VAR_texas_hosted_zone = $(TEXAS_HOSTED_ZONE)
TF_VAR_environment = $(ENVIRONMENT)
```

**Review by JackPlowman — CHANGES_REQUESTED** [review]
With these changes it should be all be working. 

A couple of small makefile changes are also required:

```
deploy: # Deploys whole project - mandatory: PROFILE
	eval "$$(make populate-deployment-variables)"
	make terraform-apply-auto-approve STACKS=lambda-security-group
	make serverless-deploy
	make terraform-apply-auto-approve STACKS=api-gateway-route53
```

``` 
undeploy: # Undeploys whole project - mandatory: PROFILE
	make serverless-remove VERSION="any"
	make terraform-destroy-auto-approve STACKS=lambda-security-group
	make terraform-destroy-auto-approve STACKS=api-gateway-route53
```

**Review by JackPlowman — CHANGES_REQUESTED** [review]
Couple of minor changes left to add. Overall looks good though

File: Makefile [hunk 12]
```diff
@@ -39,6 +40,7 @@ sls-only-deploy: # Deploys all lambdas - mandatory: PROFILE, VERSION=[commit has
 undeploy: # Undeploys whole project - mandatory: PROFILE
 	make serverless-remove VERSION="any"
 	make terraform-destroy-auto-approve STACKS=lambda-security-group
```
[the remark this answers is not in the document]
Comment (rohith1122) [reply Makefile]: Updated

File: infrastructure/stacks/api-gateway-route53/variables.tf [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+}
+variable "team_id" {
+  description = ""
+}
+variable "programme" {
+  description = ""
+}
+variable "terraform_networking_route53_zone_name" {
+    description = ""
+    default="k8s-nonprod.texasplatform.uk"
```
[the remark this answers is not in the document]
Comment (rohith1122) [reply infrastructure/stacks/api-gateway-route53/variables.tf]: Done

File: infrastructure/stacks/api-gateway-route53/variables.tf [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+# ############################
```
[the remark this answers is not in the document]
Comment (rohith1122) [reply infrastructure/stacks/api-gateway-route53/variables.tf:58]: Added

File: infrastructure/stacks/api-gateway-route53/variables.tf [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+variable "route53_terraform_state_key" {
+  description = "terraform state key"
+}
+variable "team_id" {
+  description = "team id"
+}
+variable "programme" {
+  description = "programme"
+}
+variable "terraform_networking_route53_zone_name" {
```
Comment (JackPlowman) [inline infrastructure/stacks/api-gateway-route53/variables.tf:54]: Please can this variable be replaced. Once this is done I will approve the PR.
``` suggestion
variable "environment" {
	  description = ""
	}
```

**Review by JackPlowman — APPROVED** [review]
LGTM 👍 , tested in AWS

**Comment by JackPlowman** [comment]
Sorry I found bug undeploying the application so the `deploy` make target needs to be
```
undeploy: # Undeploys whole project - mandatory: PROFILE
	make terraform-destroy-auto-approve STACKS=api-gateway-route53
	make serverless-remove VERSION="any"
	make terraform-destroy-auto-approve STACKS=lambda-security-group
	if [ "$(PROFILE)" == "task" ]; then
		make terraform-destroy-auto-approve STACKS=api-key
	fi
```

**Review by JackPlowman — APPROVED** [review]
LGTM 👍  Thanks for all the changes. Latest version has been built, deployed and undeployed.

**Review by nat-darke-nhs — APPROVED** [review]

**Pull request merged**